### PR TITLE
Multi-signal editor: part 2

### PR DESCRIPTION
**Pull request by dabreegster**
@michaelkirk: It may be easier to just look at the final code instead of trying to understand the diff, since there was some shuffling around. The changes here:

- port all the features of the old editor to the new
- get rid of the old signal editor, use the new multi-signal one even for individual signals
- proper UI for choosing multiple signals to edit

The next steps after this PR will be:

- figuring out how to draw the sidebar for intersections far apart
- the original goal: adjusting relative offset between adjacent intersections

screencast

### CODE REVIEW

File: game/src/edit/traffic_signals/picker.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            composite: Composite::new(Widget::col(vec![
+                Widget::row(vec![
+                    Line("Select multiple traffic signals")
+                        .small_heading()
+                        .draw(ctx),
+                    Btn::text_fg("X")
+                        .build(ctx, "close", hotkey(Key::Escape))
+                        .align_right(),
+                ]),
+                Btn::text_fg("Edit").build_def(ctx, hotkey(Key::E)),
```
Comment (michaelkirk) [inline game/src/edit/traffic_signals/picker.rs:37]: UX note: seeing "Edit" here confused me. Maybe something like "Continue" would be better? Take it or leave it though.

Screen Shot 2020-08-13 at 2 35 08 PM

Or if we wanted to get really fancy, something like "Edit (3) Intersections" might make it clearer that the "Edit" button takes you "back".

File: game/src/edit/traffic_signals/preview.rs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    }
+}
+
+// TODO I guess it's valid to preview without all turns possible. Some agents are just sad.
+pub fn make_previewer(
+    ctx: &mut EventCtx,
+    app: &App,
+    members: BTreeSet<IntersectionID>,
+    phase: usize,
+) -> Box<dyn State> {
```
Comment (michaelkirk) [inline game/src/edit/traffic_signals/preview.rs:77]: FYI: I got a crash when previewing random agents around the two intersections at the north east end of montlake.

```
thread 'main' panicked at 'Queue has spillover on Traversable::Lane(2093) at 00:00:01.0 -- can't draw Car #68, bound is -0.6725m. Laggy head is None. T
his is usually a geometry bug; check for duplicate roads going between the same intersections.', sim/src/mechanics/queue.rs:125:17                     
stack backtrace:                                                                                                                                       
   0: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt                                                                
   1: core::fmt::write                                                                                                                                 
   2: std::io::Write::write_fmt                                                                                                                        
   3: std::panicking::default_hook::{{closure}}                                                                                                        
   4: std::panicking::default_hook                                                                                                                     
   5: std::panicking::rust_panic_with_hook                                                                                                             
   6: rust_begin_unwind                                                                                                                                
   7: std::panicking::begin_panic_fmt                                                                                                                  
   8: sim::mechanics::queue::Queue::inner_get_car_positions                                                                                            
   9: sim::mechanics::queue::Queue::get_car_positions                                                                                                  
  10: sim::mechanics::driving::DrivingSimState::get_draw_cars_on                                                                                       
  11: <sim::sim::Sim as sim::render::GetDrawAgents>::get_draw_cars                                                                                     
  12: game::render::map::AgentCache::populate_if_needed                                                                                                
  13: game::app::App::get_renderables_back_to_front                                                                                                    
  14: game::app::App::draw                    
```

Not sure if it's related to this change or, if I'm reading the log message correctly, maybe it implies a bug in the osm source data?

File: game/src/edit/traffic_signals/picker.rs (hunk 1 above)
Comment (dabreegster) [reply game/src/edit/traffic_signals/picker.rs:37]: Done. Also made the button use the more prominent style. The fancier thing is nice -- we could even disable the button when 0 things are selected -- but I'll leave that as TODO for now.

File: game/src/edit/traffic_signals/preview.rs (hunk 2 above)
Comment (dabreegster) [reply game/src/edit/traffic_signals/preview.rs:77]: Ooh, good catch! Unrelated to this change, but I'm looking into this

File: game/src/edit/traffic_signals/mod.rs [hunk 3]
```diff
@@ -421,18 +463,101 @@ impl State for NewTrafficSignalEditor {
...
+        }
+    } else {
+        txt.add(Line(format!("{} intersections", members.len())).big_heading_plain());
+    }
+    {
+        let mut total = Duration::ZERO;
+        for p in &canonical_signal.phases {
+            total += p.phase_type.simple_duration();
+        }
+        // TODO Say "normally" to account for adaptive phases?
```
Comment (michaelkirk) [inline game/src/edit/traffic_signals/mod.rs:548]: hmmm just thinking about adaptive phases... 

If I select 4 signals to edit, and give them all adaptive phasing

Initially, they'll switch in time with eachother, but if there's traffic waiting at just one of the lights - it'll adaptively make the phase longer. Won't that cause it to get out of sync with the rest of it's edit group?

I think that's probably not a small fix and a reasonable limitation for the current system, I just wanted to see if I understand how this currently works...

File: game/src/edit/traffic_signals/edits.rs [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    mode: GameplayMode,
+    original: BundleEdits,
+) -> Box<dyn State> {
+    let has_sidewalks = app
+        .primary
+        .map
+        .get_turns_in_intersection(i)
+        .any(|t| t.between_sidewalks());
+
+    let use_template = "use template";
```
Comment (michaelkirk) [inline game/src/edit/traffic_signals/edits.rs:117]: I hadn't seen these fancy editing options before. :+1:

File: game/src/edit/traffic_signals/mod.rs (hunk 3 above)
Comment (dabreegster) [reply game/src/edit/traffic_signals/mod.rs:548]: Urgh, this is indeed a nasty case I hadn't anticipated. Originally I threw together the adaptive phase type as a proof-of-concept for https://github.com/dabreegster/abstreet/issues/91. The policy as it exists is pretty bad; a single car will repeat the entire phase for the full duration. I think there will have to be substantial future work to understand how actuated/adaptive signals work in reality, model that in abst, and keep the plethora of configuration options sane to understand in the UI.

File: game/src/edit/traffic_signals/edits.rs (hunk 4 above)
Comment (dabreegster) [reply game/src/edit/traffic_signals/edits.rs:117]: Just a few days ago, I finally cleaned up `ChooseSomething`! The API used to be this very complicated sequence-of-prompts "wizard" that kind of mimicked coroutines.

**Pull request merged**